Frameworks/Foundation/NSKVOSupport.mm

```diff
@@ -516,169 +515,301 @@ - (void)removeObserver:(id)observer forKeyPath:(NSString*)keyPath {
+                    newValue = [rootObject valueForKeyPath:keypath] ?: [NSNull null];
+                    keypathValueCache[keypath] = newValue;
+                }
 
-        NSKeyValueObservingOptions options = keypathObserver.options;
-        id rootObject = keypathObserver.object;
-        id observer = keypathObserver.observer;
-        NSString* keypath = keypathObserver.keypath;
-        NSMutableDictionary* change = keypathObserver.pendingChange;
-        void* context = keypathObserver.context;
-
-        if ((options & NSKeyValueObservingOptionNew) && [change[NSKeyValueChangeKindKey] integerValue] != NSKeyValueChangeRemoval) {
-            // For to-many mutations, we can't get the new values at indexes that have been deleted.
-            id newValue = [keypathValueCache objectForKey:keypath];
-            if (!newValue) {
-                newValue = _valueForPendingChange(notifyingObject, key, rootObject, keypath, keyObserver, change);
-                [keypathValueCache setObject:newValue forKey:keypath];
+                change[NSKeyValueChangeNewKey] = newValue;
             }
-            [change setObject:newValue forKey:NSKeyValueChangeNewKey];
-
-            // VSO 5051216: Implement set mutation notifications.
-        }
 
-        [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
+            [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
 
-        keypathObserver.pendingChange = nil;
+            keypathObserver.pendingChange = nil;
+        });
     }
 }
 
 /**
 @Status Interoperable
 */
-- (void)didChangeValueForKey:(NSString*)key {
-    if (![self observationInfo]) {
-        return;
+- (void)willChange:(NSKeyValueChange)changeKind valuesAtIndexes:(NSIndexSet*)indexes forKey:(NSString*)key {
+    if ([self observationInfo]) {
+        _dispatchWillChange(self, key, [&changeKind, indexes, key, notifyingObject = self ](_NSKVOKeyObserver * keyObserver) {
+            _NSKVOKeypathObserver* keypathObserver = keyObserver.keypathObserver;
+            NSKeyValueObservingOptions options = keypathObserver.options;
+            id rootObject = keypathObserver.object;
+            NSString* keypath = keypathObserver.keypath;
+            NSMutableDictionary* change = [NSMutableDictionary dictionaryWithDictionary:@{
+                NSKeyValueChangeKindKey : @(changeKind),
+                NSKeyValueChangeIndexesKey : indexes
+            }];
+
+            // The reference platform does not support to-many mutations on nested keypaths.
+            // We have to treat them as to-one mutations to support aggregate functions.
+            if (changeKind != NSKeyValueChangeSetting && keyObserver.restOfKeypathObserver) {
+                // This only needs to be done in willChange because didChange derives from the existing changeset.
+                change[NSKeyValueChangeKindKey] = @(changeKind = NSKeyValueChangeSetting);
```

Comment:
this has the added effect of changing `changeKind` for all future observers; is that what we want?

I don't think it is; keyObserver is local to a single observer dispatch, but changeKind if global to all observer dispatches for this key!